### PR TITLE
[XTube] Fix extraction

### PR DESCRIPTION
**Pull request by paulguy**
XTube seems to have changed their video page quite a bit again, this fixes extraction for this site.

I'm a bit rusty at python and have very little experience with this so it's more or less just me trying stuff until it works, but it seems to work.  Had to remove a few fields that couldn't easily be gleaned from the page.

### CODE REVIEW

File: youtube_dl/extractor/xtube.py [hunk 1]
```diff
@@ -51,38 +50,62 @@ def _real_extract(self, url):
...
+                continue
+            elif line[0].isalpha(): # key
+                newjsonstr += '\"' + line[0:line.index(':')] + '\":'
+                if line[line.index(':') + 2] == '\'': # string value
+                    newjsonstr += ' \"' + line[line.index(':') + 3:line.rindex('\'')]\
+                                  + '\",'
+                else:
+                    newjsonstr += line[line.index(':') + 1:]
+            else:
+                newjsonstr += line
```
Comment (dstftw) [inline youtube_dl/extractor/xtube.py]: This looks like `utils.js_to_json`.

Comment (paulguy) [inline youtube_dl/extractor/xtube.py]: It mostly works, but in this case, the site outputs a single line comment in the JSON (`//`) which _parse_json() isn't OK with.

Comment (yan12125) [inline youtube_dl/extractor/xtube.py]: Extract ```sources``` and ```videoTitle``` separately.

File: youtube_dl/extractor/xtube.py [hunk 2]
```diff
@@ -51,38 +50,44 @@ def _real_extract(self, url):
         req.add_header('Cookie', 'age_verified=1; cookiesAccepted=1')
         webpage = self._download_webpage(req, display_id)
 
-        flashvars = self._parse_json(
-            self._search_regex(
-                r'xt\.playerOps\s*=\s*({.+?});', webpage, 'player ops'),
-            video_id)['flashvars']
+#        jsonstr = self._search_regex(
+#                r'MHP1138\.createPlayer\(\'xtubePlayer\',\s*{.*medias:\s*({.+?}),\s*menu:',
+#                webpage, 'player ops', flags=re.DOTALL, group=1)
```
Comment (yan12125) [inline youtube_dl/extractor/xtube.py]: Remove unused codes.

File: youtube_dl/extractor/xtube.py [hunk 3]
```diff
@@ -51,38 +50,44 @@ def _real_extract(self, url):
...
 
-        title = flashvars.get('title') or self._search_regex(
-            r'<h1>([^<]+)</h1>', webpage, 'title')
-        video_url = compat_urllib_parse_unquote(flashvars['video_url'])
-        duration = int_or_none(flashvars.get('video_duration'))
+        sources = self._search_regex(r'sources:\s*({.+?}),', webpage, 'sources')
+
+        sourcesdict = self._parse_json(sources, video_id)
+        print(str(sourcesdict))
```
Comment (yan12125) [inline youtube_dl/extractor/xtube.py]: Remove debugging codes.

File: youtube_dl/extractor/xtube.py [hunk 4]
```diff
@@ -51,38 +50,44 @@ def _real_extract(self, url):
...
+
+        formats = []
+        for formatkey in sourcesdict.keys():
+            curformat = {}
+            curformat['height'] = int(formatkey)
+            curformat['url'] = sourcesdict[formatkey]
+            formats.append(curformat)
+            
+        formats = sorted(formats, key=lambda x: x['height'])
```
Comment (yan12125) [inline youtube_dl/extractor/xtube.py]: Use ```self._sort_formats(formats)``` instead.

File: youtube_dl/extractor/xtube.py (hunk 2 above)
Comment (paulguy) [inline youtube_dl/extractor/xtube.py]: Oops.

**Comment by yan12125** [comment]
Thanks. Could you squash all commits into a single one?

**Comment by paulguy** [comment]
I'm not sure how...

**Comment by yan12125** [comment]
Well, the easiest way is to re-commit changes.

File: youtube_dl/extractor/xtube.py [hunk 5]
```diff
@@ -51,38 +50,39 @@ def _real_extract(self, url):
...
 
         return {
             'id': video_id,
             'display_id': display_id,
-            'url': video_url,
+            'formats': formats,
             'title': title,
             'description': description,
             'uploader': uploader,
-            'duration': duration,
```
Comment (dstftw) [inline youtube_dl/extractor/xtube.py]: It's available and should be extracted.

File: youtube_dl/extractor/xtube.py [hunk 6]
```diff
@@ -51,38 +50,39 @@ def _real_extract(self, url):
...
             'view_count': view_count,
-            'comment_count': comment_count,
```
Comment (dstftw) [inline youtube_dl/extractor/xtube.py]: This as well.

File: youtube_dl/extractor/xtube.py [hunk 7]
```diff
@@ -51,38 +50,39 @@ def _real_extract(self, url):
...
-        title = flashvars.get('title') or self._search_regex(
-            r'<h1>([^<]+)</h1>', webpage, 'title')
-        video_url = compat_urllib_parse_unquote(flashvars['video_url'])
-        duration = int_or_none(flashvars.get('video_duration'))
+        sourcesdict = self._parse_json(sources, video_id)
+
+        formats = []
+        for formatkey in sourcesdict.keys():
+            curformat = {}
+            curformat['height'] = int(formatkey)
```
Comment (dstftw) [inline youtube_dl/extractor/xtube.py]: Use `int_or_none` instead.

File: youtube_dl/extractor/xtube.py [hunk 8]
```diff
@@ -51,38 +50,39 @@ def _real_extract(self, url):
...
+        sources = self._search_regex(r'sources:\s*({.+?}),', webpage, 'sources')
 
-        title = flashvars.get('title') or self._search_regex(
-            r'<h1>([^<]+)</h1>', webpage, 'title')
-        video_url = compat_urllib_parse_unquote(flashvars['video_url'])
-        duration = int_or_none(flashvars.get('video_duration'))
+        sourcesdict = self._parse_json(sources, video_id)
+
+        formats = []
+        for formatkey in sourcesdict.keys():
```
Comment (dstftw) [inline youtube_dl/extractor/xtube.py]: Use `for key, value in sourcesdict.items()` instead.

**Comment by paulguy** [comment]
I guess I'll just leave it up to people who know what they're doing next time.

**Pull request closed without merging**